crates/driver/src/domain/competition/solution/settlement.rs

```diff
@@ -133,6 +143,75 @@ impl Verified {
     pub fn boundary(self) -> boundary::Settlement {
         self.inner.boundary
     }
+
+    /// The solutions encoded in this settlement. This is a [`HashSet`] because
+    /// multiple solutions can be encoded in a single settlement due to
+    /// merging. See [`Verified::merge`].
+    pub fn solutions(&self) -> &HashSet<super::Id> {
+        &self.inner.solutions
+    }
+
+    /// Merge another settlement into this settlement.
+    ///
+    /// Merging settlements results in a score that can be anything due to the
+    /// fact that contracts can do basically anything, but in practice it can be
+    /// assumed that the score will be at least equal to the sum of the scores
+    /// of the merged settlements.
+    pub async fn merge(
+        &self,
+        other: &Self,
+        contract: &contracts::GPv2Settlement,
+        eth: &Ethereum,
+        simulator: &Simulator,
+    ) -> Result<Self, super::Error> {
+        // The solver must be the same for both settlements.
+        if self.inner.boundary.solver != other.inner.boundary.solver {
+            return Err(super::Error::DifferentSolvers);
+        }
+
+        // If the solutions being merged are not disjoint, the settlements can't be
+        // merged.
+        if !self.inner.solutions.is_disjoint(&other.inner.solutions) {
+            return Err(super::Error::DuplicateSolutions);
+        }
+
+        let settlement = self
+            .inner
+            .boundary
+            .clone()
+            .merge(other.inner.boundary.clone())?;
+
+        let gas = Gas {
+            // After two settlements have been merged, the only verification that
+            // needs to be done is simulating the settlement (which also yields the gas
+            // needed by the final merged settlement). All other verified rules still
+            // hold after the settlements have been merged. See also [`Solution::verify`].
+            estimate: simulator
+                .gas(settlement.tx(contract, Internalization::Enable))
+                .await?,
+            ..self.gas
+        };
+
+        // Ensure that the solver has sufficient balance for the settlement to be mined.
+        if eth.balance(self.inner.solver()).await? < gas.required_balance() {
+            return Err(super::Error::InsufficientBalance);
+        }
+
+        Ok(Self {
+            inner: Settlement {
+                id: self.inner.id,
+                solutions: self
+                    .inner
+                    .solutions
+                    .union(&other.inner.solutions)
+                    .copied()
+                    .collect(),
+                boundary: settlement,
+            },
+            access_list: self.access_list.clone().merge(other.access_list.clone()),
```

Comment:
It might not be enough to merge the access lists. Merging multiple settlements might result in different storage slots being read than in any of the individual settlements.
Not sure how often this would be a problem, though.

Reply:
https://github.com/cowprotocol/services/pull/1490/commits/8c0d5f3549b5455cb1c08f98989bf8a332963d1a